alf/networks/transformer_networks.py

```diff
@@ -222,3 +227,167 @@ def forward(self, inputs, state=()):
+            last_layer_size (None): nt used; for interface compatibility
+            last_activation (None): not used; for interface compatibility
+            last_kernel_initializer (None): not used; for interface compatibility
+            last_use_fc_bn (None): not used; for interface compatibility
+            name (str):
+        """
+        super().__init__(
+            input_tensor_spec,
+            input_preprocessors,
+            preprocessing_combiner=preprocessing_combiner,
+            name=name)
+
+        if kernel_initializer is None:
+            kernel_initializer = functools.partial(
+                variance_scaling_init,
+                mode='fan_in',
+                distribution='truncated_normal',
+                nonlinearity=activation)
+
+        embedding_layers = nn.ModuleList()
+        print(self._processed_input_tensor_spec)
+        assert self._processed_input_tensor_spec.ndim == 2, (
+            "expect the "
+            "processed spec to have the shape of [entity_num, feature_dim]")
+
+        for size in fc_layer_params:
+            embedding_layers.append(
+                layers.FC(
+                    self._processed_input_tensor_spec.shape[-1],
+                    size,
+                    activation=activation,
+                    use_bn=use_fc_bn,
+                    kernel_initializer=kernel_initializer))
+            input_size = size
+        self._embedding_layers = embedding_layers
+
+        # input_size = N * d'
+        full_feature_dim = input_size
+        fea_dim = input_size
+        assert fea_dim % num_of_heads == 0, "improper value for num_of_heads"
+        self._num_of_heads = num_of_heads
+        self._fea_dim_per_head = fea_dim // num_of_heads
+
+        # attention related layers
+        self._value_proj = layers.FC(
+            full_feature_dim,
+            fea_dim,
+            use_bias=False,
+            kernel_initializer=kernel_initializer)
+        self._key_proj = layers.FC(
+            full_feature_dim,
+            fea_dim,
+            use_bias=False,
+            kernel_initializer=kernel_initializer)
+        self._query_proj = layers.FC(
+            full_feature_dim,
+            fea_dim,
+            use_bias=False,
+            kernel_initializer=kernel_initializer)
+
+        self._attention_res_branch = layers.FC(
+            fea_dim,
+            fea_dim,
+            use_bias=False,
+            kernel_initializer=kernel_initializer)
+
+        self._mha = alf.layers.MultiHeadAttention()
+
+    def forward(self, inputs, state=()):
+        """
+        Args:
+            inputs (Tensor):  with the shape of [B, N, d], where
+                B denotes batch size, N the number of entities, and d the
+                feature dimension
+            state (nested Tensor): states
+        Returns:
+            - Tensor: shape is [B, k], where k denotes the dimension of output
```

Comment:
comment what the value of k is

Reply:
It denotes the dimensionality of the output feature.  Updated the comments.